python/cudf/cudf/core/internals/where.py

```diff
@@ -0,0 +1,375 @@
+import warnings
+
+import numpy as np
+import pandas as pd
+
+import cudf
+
+
+def _normalize_scalars(col, other):
+    """
+    Try to normalizes scalar values as per col dtype
+    """
+    if (
+        other is not None
+        and (isinstance(other, float) and not np.isnan(other))
+    ) and (col.dtype.type(other) != other):
+        raise TypeError(
+            f"Cannot safely cast non-equivalent "
+            f"{type(other).__name__} to {col.dtype.name}"
+        )
+
+    return (
+        col.dtype.type(other)
+        if (
+            other is not None
+            and (isinstance(other, float) and not np.isnan(other))
+        )
+        else other
+    )
+
+
+def _check_and_cast_columns(source_col, other_col, inplace):
+    """
+    Returns type-casted columns of `source_col` & `other_col`
+    based on `inplace` parameter.
+    """
+    if cudf.utils.dtypes.is_categorical_dtype(source_col.dtype):
+        return source_col, other_col
+    elif cudf.utils.dtypes.is_mixed_with_object_dtype(source_col, other_col):
+        raise TypeError(
+            "cudf does not support mixed types, please type-cast "
+            "the column of dataframe/series and other "
+            "to same dtypes."
+        )
+    if inplace:
+        if not source_col.can_cast_safely(other_col.dtype):
+            warnings.warn(
+                f"Type-casting from {other_col.dtype} "
+                f"to {source_col.dtype}, there could be potential data loss"
+            )
+        return source_col, other_col.astype(source_col.dtype)
+    else:
+        common_dtype = cudf.utils.dtypes.find_common_type(
+            [source_col.dtype, other_col.dtype]
+        )
+        return source_col.astype(common_dtype), other_col.astype(common_dtype)
+
+
+def _check_and_cast_columns_with_scalar(source_col, other_scalar, inplace):
+    """
+    Returns type-casted column `source_col` & scalar `other_scalar`
+    based on `inplace` parameter.
+    """
+    if cudf.utils.dtypes.is_categorical_dtype(source_col.dtype):
+        return source_col, other_scalar
+
+    device_scalar = cudf.Scalar(
+        _normalize_scalars(source_col, other_scalar),
+        dtype=source_col.dtype if other_scalar is None else None,
+    )
+
+    if other_scalar is None:
+        return source_col, device_scalar
+    elif cudf.utils.dtypes.is_mixed_with_object_dtype(
+        device_scalar, source_col
+    ):
+        raise TypeError(
+            "cudf does not support mixed types, please type-cast "
+            "the column of dataframe/series and other "
+            "to same dtypes."
+        )
+    if inplace:
+        if not np.can_cast(device_scalar, source_col.dtype):
```

Comment:
Is it safe for us to use `np.can_cast` here? We should be handling a `DecimalDtype` which numpy won't know how to handle

Reply:
I have created a new utility named `cudf.utils.dtypes.can_cast` that should now handle our dtypes without erroring when sent to `np.can_cast`.